plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+    TyLam _ x ka b    -> [ (KindArrow () ka' kb, TyLam () x ka' $ substClosedType x (minimalType ka) b)
+                         | ka' <- shrink ka] ++
+                         [ (KindArrow () ka kb', TyLam () x ka b)
+                         | (kb', b) <- shrinkKindAndType (Map.insert x ka ctx) (kb, b)]
+      where KindArrow _ _ kb = k
+    TyForall _ x ka b -> [ (k, b) | not $ Set.member x (fvType b) ] ++
+                         -- (above) If the bound variable doesn't matter we get rid of the binding
+                         [ (k, TyForall () x ka' $ substClosedType x (minimalType ka) b)
+                         | ka' <- shrink ka] ++
+                         -- (above) we can always just shrink the bound variable to a smaller kind
+                         -- and ignore it
+                         [ (k, TyForall () x ka b)
+                         | (_, b) <- shrinkKindAndType (Map.insert x ka ctx) (Star, b)]
+                         -- (above) or we shrink the body
+    TyBuiltin{}       -> []
+    TyIFix{}          -> error "shrinkKindAndType: TyIFix"
+
+-- CODE REVIEW: does this exist anywhere?
+inferKind :: Map TyName (Kind ()) -> Type TyName DefaultUni () -> Maybe (Kind ())
+inferKind ctx ty = case ty of
+  TyVar _ x        -> Map.lookup x ctx
+  TyFun _ _ _      -> pure $ Star
+  TyApp _ a _      -> do KindArrow _ _ k <- inferKind ctx a; pure k
+  TyLam _ x k b    -> KindArrow () k <$> inferKind (Map.insert x k ctx) b
+  TyForall _ _ _ _ -> pure $ Star
+  TyBuiltin _ b    -> pure $ builtinKind b
+  TyIFix{}         -> error "inferKind: TyIFix"
+
+-- | Partial inferKind_, useful for context where invariants are set up to guarantee
+-- that types are well-kinded.
+inferKind_ :: HasCallStack => Map TyName (Kind ()) -> Type TyName DefaultUni () -> Kind ()
+inferKind_ ctx ty =
+  case inferKind ctx ty of
+    Nothing -> error "inferKind"
+    Just k  -> k
+
+-- | Shrink a type in a context assuming that it is of kind *.
+shrinkType :: HasCallStack
+           => Map TyName (Kind ())
+           -> Type TyName DefaultUni ()
+           -> [Type TyName DefaultUni ()]
+shrinkType ctx ty = map snd $ shrinkKindAndType ctx (Star, ty)
+
+-- | Shrink a type of a given kind in a given context in a way that keeps its kind
+shrinkTypeAtKind :: HasCallStack
+                 => Map TyName (Kind ())
+                 -> Kind ()
+                 -> Type TyName DefaultUni ()
+                 -> [Type TyName DefaultUni ()]
+shrinkTypeAtKind ctx k ty = [ ty' | (k', ty') <- shrinkKindAndType ctx (k, ty), k == k' ]
+
+data Polarity = Pos
+              | Neg
+              deriving stock (Ord, Eq, Show)
+
+substType :: HasCallStack
+          => Map TyName (Type TyName DefaultUni ())
+          -> Type TyName DefaultUni ()
+          -> Type TyName DefaultUni ()
+substType = substType' True
+
+-- | Generalized substitution algorithm
+substType' :: HasCallStack
```

Comment:
If you've got a bunch of generalized substitution stuff I suggest pulling it out into a standalone thing. It think we may not exactly have a naive type substitution function like this because the one we use in the typechecker does normalizing substitution: https://github.com/input-output-hk/plutus/blob/master/plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs#L199